### DISCUSSION TITLE
The use of globals in TinyUSB

### DISCUSSION THREAD

**Discussion by mikemoretti3** [hathach/tinyusb#773] [General]
Hi,

First off, I really like TinyUSB.

However, the architecture makes it somewhat difficult to not have to hack globals for stuff everywhere.  It would've been so much better if it was a little more "object-oriented".  E.g. maybe have a "context" with callback function pointers (that take the context as a parameter) and user data for each interface, and each configuration.  It would make it easier to design an app that say has two cdc interfaces (or more than one of any class interface for that matter) that do something completely different and really should be separated code-wise, whereas right now there's one class task per class that has to loop through interface numbers and choose what to do.

Am I the only one that sees this as an issue?  Maybe I'm looking at it wrong?

-m

**Comment by hathach** [comment]
Since I read C better than English, can you elaborate your issue with the actual project code
1. what are you trying to do e.g with multiple CDCs, and how the APIs cause the issue
2. do you have any suggestion in mind. Please try to make an PR for it, I am open to all the changes.

**Reply by mikemoretti3** [reply]
Unfortunately, what I'm suggesting would probably require drastic changes to the API.  I don't know if it would be a good idea at the current stage of the project to make those kinds of changes.

**Reply by hathach** [reply]
major changes is not an problem if it does fix any real issue of the existing API. What is your problem with existing API in your project. Can you post your code with description for the issue.
